### PR TITLE
Implement a new algorithm for the CUDA plugin

### PR DESCRIPTION
**Pull request by YuanTingHsieh**
The original algorithm looping through every bin and for each bin calls the reduce_sum on all the samples that belong to this bin, the number of total samples denoted as N.
The total number of bins can be calculated by F x B.
When the total number of bins  (FB) is huge, the original algorithm will launch F x B CUDA kernels, which is very slow.
The complexity is roughly FB log(N) + FB*C (C is the overhead to launch a CUDA kernel)
So we need to enhance it.

The new algorithm will first layout all the numbers to be added, and each kernel launch will reduce the number by half.
Thus we only need to launch kernel log(F x N) times.
The log(FxN) is much smaller then F x B.

### Description

- Implement a new algorithm for the CUDA plugin
- Enhance the LocalPlugin interface to reduce unnecessary conversion for the new CUDA plugin

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

File: integration/xgboost/encryption_plugins/v.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
+#!/usr/bin/python
```
Comment (ZiyueXu77) [inline integration/xgboost/encryption_plugins/v.py:1]: Add license header

File: integration/xgboost/encryption_plugins/cuda_plugin/src/cuda_utils.h [hunk 2]
```diff
@@ -45,7 +46,6 @@ const static unsigned int key_len=1024;
 //const static unsigned int bits=6144;
 //const static unsigned int key_len=3072;
 
-
 const static int TPB=512;
```
Comment (ZiyueXu77) [inline integration/xgboost/encryption_plugins/cuda_plugin/src/cuda_utils.h:49]: are these two needed (not sure their meaning)? maybe also remove the unused lines

Comment (ZiyueXu77) [reply integration/xgboost/encryption_plugins/cuda_plugin/src/cuda_utils.h:49]: ah I see they are used later in paillier.h, maybe move to there if that's the only place using this?

File: integration/xgboost/encryption_plugins/v.py (hunk 1 above)
Comment (ZiyueXu77) [reply integration/xgboost/encryption_plugins/v.py:1]: for this testing script, maybe we move it to the example folder?

**Review by ZiyueXu77 — COMMENTED** [review]
some initial comments

File: integration/xgboost/encryption_plugins/v.py (hunk 1 above)
Comment (YuanTingHsieh) [reply integration/xgboost/encryption_plugins/v.py:1]: yes, I should remove it

File: integration/xgboost/encryption_plugins/cuda_plugin/src/cuda_utils.h (hunk 2 above)
Comment (YuanTingHsieh) [reply integration/xgboost/encryption_plugins/cuda_plugin/src/cuda_utils.h:49]: it's also used in cuda_plugin.h, so keep it here for now.

File: integration/xgboost/encryption_plugins/cuda_plugin/src/cuda_plugin.h [hunk 3]
```diff
@@ -56,16 +120,28 @@ class CUDAPlugin: public LocalPlugin {
...
 
     void setGHPairs() {
       if (debug_) std::cout << "setGHPairs is called" << std::endl;
       const std::uint8_t* pointer = encrypted_gh_.data();
 
       // Retrieve header
-      BufferHeader header;
-      std::memcpy(&header, pointer, sizeof(BufferHeader));
-      pointer += sizeof(BufferHeader);
+      BufferHeader2 header;
```
Comment (nvidianz) [inline integration/xgboost/encryption_plugins/cuda_plugin/src/cuda_plugin.h:142]: Should just use BufferHeader

Comment (YuanTingHsieh) [reply integration/xgboost/encryption_plugins/cuda_plugin/src/cuda_plugin.h:142]: good catch! will fix

**Comment by yanchengnv** [comment]
@nvidianz and @YuanTingHsieh please get this PR merged before YT goes to TW.

**Review by nvidianz — APPROVED** [review]
The new names are much better.

**Pull request merged**